pkg/pgmodel/sql_ingest.go

```diff
@@ -465,31 +465,85 @@ func (h *insertHandler) flushPending() {
 }
 
 func runInserter(conn pgxConn, in chan copyRequest) {
+	insertBatch := make([]copyRequest, 0, 100)
 	for {
-		req, ok := <-in
+		var ok bool
+		insertBatch, ok = inserterGetBatch(insertBatch, in)
 		if !ok {
 			return
 		}
-		err := doInsert(conn, req)
+
+		// sort to prevent deadlocks
+		sort.Slice(insertBatch, func(i, j int) bool {
+			return insertBatch[i].table < insertBatch[j].table
+		})
+
+		numInsertsPerBatch.Observe(float64(len(insertBatch)))
+
+		doInsertOrFallback(conn, insertBatch...)
+		for i := range insertBatch {
+			insertBatch[i] = copyRequest{}
+		}
+		insertBatch = insertBatch[:0]
+	}
+}
+
+func inserterGetBatch(batch []copyRequest, in chan copyRequest) ([]copyRequest, bool) {
+	req, ok := <-in
+	if !ok {
+		return batch, false
+	}
+	batch = append(batch, req)
+	timeout := time.After(100 * time.Millisecond)
+hot_gather:
+	for len(batch) < cap(batch) {
+		select {
+		case r2 := <-in:
+			batch = append(batch, r2)
+		case _ = <-timeout:
+			break hot_gather
+		}
+	}
+	return batch, true
+}
+
+func doInsertOrFallback(conn pgxConn, reqs ...copyRequest) {
+	err := doInsert(conn, reqs...)
+	if err != nil {
+		insertBatchErrorFallback(conn, reqs...)
+		return
+	}
+
+	for i := range reqs {
```

Comment:
Why not directly reference the req in reqs, instead of using index. Same for `insertBatchErrorFallback()`.